# Capstone Project Part 4: Combining & Cleaning DataFrames

**Authur:** Kate Meredith  

**Date:** September-November 2022

**Notebook #:** 4 of

## Background

**Source:** Data was collected from [CoffeeReview.com](https://www.coffeereview.com/) and grouped into two DataFrames for cleaning. This workbook compiles those two datasets and further cleans them.

**Goal:** Combine the two datasets into one.

**Data Overview** Combined final columns to be as follows:

Independent Variables:
- coffee_name
- roaster_name
- month
- year
- bean_agtron
- ground_agtron
- aroma
- acidity
- body
- flavor
- aftertaste
- with_milk
- roaster_lat
- roaster_lon
- origin_lat
- origin_lon
- p1
- p2
- p3

Target: 
- Overall Score

## References

- Used this [article](https://www.geeksforgeeks.org/change-the-order-of-a-pandas-dataframe-columns-in-python/) to reorder columns
- Used this [article](https://stackoverflow.com/questions/17071871/how-do-i-select-rows-from-a-dataframe-based-on-- column-values) to figure out how to find a row based on specific value
- Checking which rows are missing values through this [help](https://www.geeksforgeeks.org/working-with-missing-data-in-pandas/)
- Dropping rows based on [column missing values](https://stackoverflow.com/questions/13413590/how-to-drop-rows-of-pandas-dataframe-whose-value-in-a-certain-column-is-nan)

In [1]:
#Importing libraries.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Combining Datasets

The following combines the two data sets with some priliminary adjustment.

In [2]:
#import first dataset
df_1 = pd.read_csv('1st_clean_data_df.csv')

In [3]:
#import second dataset
df_2 = pd.read_csv('2nd_clean_data_df.csv')

In [4]:
#initial review of data types, missing values
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4779 entries, 0 to 4778
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   coffee_name           4779 non-null   object 
 1   roaster_name          4779 non-null   object 
 2   roast_level           4705 non-null   float64
 3   overall_score         4779 non-null   int64  
 4   p1                    4779 non-null   object 
 5   p2                    4779 non-null   object 
 6   p3                    4773 non-null   object 
 7   bean_agtron           4742 non-null   float64
 8   ground_agtron         4743 non-null   float64
 9   year                  4779 non-null   int64  
 10  month                 4779 non-null   int64  
 11  roaster_location_lat  4738 non-null   float64
 12  roaster_location_lon  4738 non-null   float64
 13  origin_lat            4513 non-null   float64
 14  origin_lon            4513 non-null   float64
 15  aroma                

In [5]:
#initial review of data types, missing values
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1788 entries, 0 to 1787
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   coffee_name           1788 non-null   object 
 1   roaster_name          1788 non-null   object 
 2   overall_score         1784 non-null   float64
 3   p1                    1788 non-null   object 
 4   p2                    1788 non-null   object 
 5   p3                    1788 non-null   object 
 6   roast                 1510 non-null   float64
 7   aroma                 1785 non-null   float64
 8   acidity               1452 non-null   float64
 9   body                  1733 non-null   float64
 10  flavor                767 non-null    float64
 11  aftertaste            2 non-null      float64
 12  with_milk             9 non-null      float64
 13  bean_agtron           1549 non-null   float64
 14  ground_agtron         1554 non-null   float64
 15  year                 

In [6]:
#update column name to match other df
df_1.rename(columns = {'roast_level':'roast'}, inplace = True)

In [7]:
#reorder columns
df_1 = df_1[['coffee_name', 'roaster_name', 'month', 'year', 'roast', 'bean_agtron', 'ground_agtron', 'aroma', 'acidity', 'body', 'flavor', 'aftertaste', 'with_milk', 'roaster_location_lat', 'roaster_location_lon', 'origin_lat', 'origin_lon', 'p1', 'p2', 'p3', 'overall_score']]

In [8]:
#checking order update
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4779 entries, 0 to 4778
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   coffee_name           4779 non-null   object 
 1   roaster_name          4779 non-null   object 
 2   month                 4779 non-null   int64  
 3   year                  4779 non-null   int64  
 4   roast                 4705 non-null   float64
 5   bean_agtron           4742 non-null   float64
 6   ground_agtron         4743 non-null   float64
 7   aroma                 4743 non-null   float64
 8   acidity               4049 non-null   float64
 9   body                  4779 non-null   int64  
 10  flavor                4779 non-null   int64  
 11  aftertaste            4779 non-null   int64  
 12  with_milk             766 non-null    float64
 13  roaster_location_lat  4738 non-null   float64
 14  roaster_location_lon  4738 non-null   float64
 15  origin_lat           

In [9]:
#redording columns
df_2 = df_2[['coffee_name', 'roaster_name', 'month', 'year', 'roast', 'bean_agtron', 'ground_agtron', 'aroma', 'acidity', 'body', 'flavor', 'aftertaste', 'with_milk', 'roaster_location_lat', 'roaster_location_lon', 'origin_lat', 'origin_lon', 'p1', 'p2', 'p3', 'overall_score']]

In [10]:
#checking order update
df_2.head()

coffee_name                 roaster_name  month  year  \
0      Doi Chaang Wild Civet Coffee            Doi Chaang Coffee      6  2009   
1                   Kenya AA Lenana                 Après Coffee      6  2009   
2             Mele 100% Kona Coffee                   Hula Daddy      6  2009   
3  Kenya Peaberry Thika Gethumbwini          JBC Coffee Roasters      6  2009   
4             Kenya Gititu Peaberry  Atomic Cafe Coffee Roasters      6  2009   

   roast  bean_agtron  ground_agtron  aroma  acidity  body  ...  aftertaste  \
0    2.0         49.0           80.0    8.0      8.0   7.0  ...         NaN   
1    4.0         42.0           53.0    8.0      8.0   7.0  ...         NaN   
2    2.0         51.0           73.0    8.0      8.0   7.0  ...         NaN   
3    1.0         57.0           90.0    8.0      9.0   8.0  ...         NaN   
4    3.0         44.0           63.0    8.0      8.0   8.0  ...         NaN   

   with_milk  roaster_location_lat  roaster_location_lon  origin_lat  \
0        NaN             51.046095           -114.065465   14.897192   
1        NaN             40.038130            -76.305669    1.441968   
2        NaN             19.623815           -155.953638   19.593802   
3        NaN             43.074761            -89.383761    1.441968   
4        NaN             42.558428            -70.880049    1.441968   

   origin_lon                                                 p1  \
0  100.832730  Blind Assessment: Intriguing mid-tones through...   
1   38.431398  Blind Assessment: Rich, very intense aroma: da...   
2 -155.428370  Blind Assessment: An exciting, rather unusual ...   
3   38.431398  Blind Assessment: Clean, complex, impeccable. ...   
4   38.431398  Blind Assessment: Rich, complex fruit aroma, i...   

                                                  p2  \
0  Notes: Doi Chaang is a single-estate coffee pr...   
1  Notes: Despite stresses brought on by social u...   
2  Notes: A blend of coffees processed by a varie...   
3  Notes: Despite national coffee leadership mark...   
4  Notes: Despite stresses brought on by social u...   

                                                  p3 overall_score  
0  Who Should Drink It: Culinary adventurers who ...          90.0  
1  Who Should Drink It: Those who prefer understa...          92.0  
2  Who Should Drink It: An exhilarating sensory r...          92.0  
3  Who Should Drink It: Strikingly complete expre...          96.0  
4  Who Should Drink It: Lovers of sweet fruit fla...          92.0  

[5 rows x 21 columns]

In [11]:
df_1.shape

(4779, 21)

df_1 has 4,779 rows and 21 columns

In [12]:
df_2.shape

(1788, 21)

df_2 has 1,788 rows and 21 columns

In [13]:
#combining dataframes
coffee = pd.concat([df_1, df_2], axis=0)

## EDA and Further Cleaning

Exploratory data analysis for combined dataframe `coffee`:

In [14]:
#checking new dataframe shape
coffee.shape

(6567, 21)

Full dataframe has 6,567 rows and 21 columns

### Updating datatypes

In [15]:
#updating to integers
coffee['roast'] = coffee['roast'].astype('Int8')

In [16]:
#updating datatypes to integers
coffee['bean_agtron'] = coffee['bean_agtron'].astype('Int8')

coffee['ground_agtron'] = coffee['ground_agtron'].astype('Int8')

coffee['aroma'] = coffee['aroma'].astype('Int8')

coffee['acidity'] = coffee['acidity'].astype('Int8')

coffee['body'] = coffee['body'].astype('Int8')

coffee['flavor'] = coffee['flavor'].astype('Int8')

coffee['aftertaste'] = coffee['aftertaste'].astype('Int8')

coffee['with_milk'] = coffee['with_milk'].astype('Int8')

coffee['overall_score'] = coffee['overall_score'].astype('Int8')

In [17]:
coffee.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6567 entries, 0 to 1787
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   coffee_name           6567 non-null   object 
 1   roaster_name          6567 non-null   object 
 2   month                 6567 non-null   int64  
 3   year                  6567 non-null   int64  
 4   roast                 6215 non-null   Int8   
 5   bean_agtron           6291 non-null   Int8   
 6   ground_agtron         6297 non-null   Int8   
 7   aroma                 6528 non-null   Int8   
 8   acidity               5501 non-null   Int8   
 9   body                  6512 non-null   Int8   
 10  flavor                5546 non-null   Int8   
 11  aftertaste            4781 non-null   Int8   
 12  with_milk             775 non-null    Int8   
 13  roaster_location_lat  6495 non-null   float64
 14  roaster_location_lon  6495 non-null   float64
 15  origin_lat           

In [18]:
#previewing data
coffee.head()

coffee_name     roaster_name  month  year  roast  \
0              Colombia Cerro Azul Enano  Equator Coffees     10  2022      2   
1                         Peru Incahuasi     Press Coffee     10  2022      2   
2            Colombia Aponte’s Guardians     Press Coffee     10  2022      2   
3        Nicaragua Flor de Dalia Natural  Equator Coffees     10  2022      2   
4  Ethiopia Bench Maji Geisha G1 Natural  Taster's Coffee     10  2022      1   

   bean_agtron  ground_agtron  aroma  acidity  body  ...  aftertaste  \
0           60             77      9        9     9  ...           8   
1           58             78      9        9     9  ...           8   
2           59             77      9        9     8  ...           8   
3           62             78      8        8     9  ...           8   
4           65             81      9        9     8  ...           8   

   with_milk  roaster_location_lat  roaster_location_lon  origin_lat  \
0       <NA>             37.973535           -122.531087    4.099917   
1       <NA>             33.448437           -112.074141   -6.869970   
2       <NA>             33.448437           -112.074141    4.099917   
3       <NA>             37.973535           -122.531087   12.609016   
4       <NA>             25.072134            121.679919   10.211670   

   origin_lon                                                 p1  \
0  -72.908813  Blind Assessment: Elegantly fruit- and cocoa-t...   
1  -75.045851  Blind Assessment: Gently fruit-toned, integrat...   
2  -72.908813  Blind Assessment: Richly sweet, spice-toned. L...   
3  -85.293691  Blind Assessment: Gently fruit-forward, sweetl...   
4   38.652120  Blind Assessment: Gently sweet-tart, floral-to...   

                                                  p2  \
0  Notes: Produced at Finca Cerro Azul (also owne...   
1  Notes: Produced at Incahuasi Farm from trees o...   
2  Notes: Produced at Aponte Farm from an undiscl...   
3  Notes: Produced by smallholding members of the...   
4  Notes: Produced from trees of the admired bota...   

                                                  p3 overall_score  
0  The Bottom Line:  This rare Enano (dwarf Geish...            94  
1  The Bottom Line: Laden with tropical fruit not...            94  
2  The Bottom Line: A balanced, inviting washed C...            93  
3  The Bottom Line: A refreshing, very sweet natu...            92  
4  The Bottom Line: A quietly confident natural-p...            93  

[5 rows x 21 columns]

In [19]:
coffee.tail()

coffee_name           roaster_name  month  year  roast  \
1783       Ethiopia Sidamo        Caravali Coffee      3  1997      5   
1784      Bridgetown Blend      Bridgetown Coffee      2  1997      4   
1785       Breakfast Blend  Green Mountain Coffee      2  1997      3   
1786  Seattle’s Best Blend  Seattle's Best Coffee      2  1997      4   
1787       Colombian Blend    Thanksgiving Coffee      2  1997      4   

      bean_agtron  ground_agtron  aroma  acidity  body  ...  aftertaste  \
1783           36             39      6        5     6  ...        <NA>   
1784           47             52      8        6     6  ...        <NA>   
1785           56             65      7        8     7  ...        <NA>   
1786           47             52      7        6     8  ...        <NA>   
1787           45             49      7        4     5  ...        <NA>   

      with_milk  roaster_location_lat  roaster_location_lon  origin_lat  \
1783       <NA>             47.382690           -122.227027   44.933143   
1784       <NA>             45.520247           -122.674194   44.933143   
1785       <NA>             44.337125            -72.755497   44.933143   
1786       <NA>             47.603832           -122.330062   -2.483383   
1787       <NA>             39.445723           -123.805294    4.099917   

      origin_lon                                                 p1  \
1783    7.540121  Blind Assessment: Pungent, chocolate notes enl...   
1784    7.540121  Blind Assessment: An intense coffee, particula...   
1785    7.540121  Blind Assessment: An exquisitely balanced trad...   
1786  117.890285  Blind Assessment: This coffee gets better as i...   
1787  -72.908813  Blind Assessment: This coffee is so complex at...   

                                                     p2  \
1783  Notes: Pungent, chocolate notes enliven the ar...   
1784  Notes: Intense coffee where carbon tones domin...   
1785  Notes: An exquisitely balanced traditional Ame...   
1786  Notes: Coffee with a heavy body that get bette...   
1787  Notes: A coffee so complex at the top that it ...   

                                                     p3 overall_score  
1783  Who Should Drink It: People who fondly remembe...            74  
1784  Who Should Drink It: People who like everythin...            77  
1785  Who Should Drink It: Traditionalists who are a...            90  
1786  Who Should Drink It: Classicists who brew with...            88  
1787  Who Should Drink It: Romantics. Should do well...            85  

[5 rows x 21 columns]

### Updating column names to be shorter

In [20]:
#shortening roaster lat and lon names
coffee.rename(columns = {'roaster_location_lat':'roaster_lat'}, inplace = True)
coffee.rename(columns = {'roaster_location_lon':'roaster_lon'}, inplace = True)

### Investigating null values:

In [21]:
coffee.isna().sum()

coffee_name         0
roaster_name        0
month               0
year                0
roast             352
bean_agtron       276
ground_agtron     270
aroma              39
acidity          1066
body               55
flavor           1021
aftertaste       1786
with_milk        5792
roaster_lat        72
roaster_lon        72
origin_lat        589
origin_lon        589
p1                  0
p2                  0
p3                  6
overall_score       4
dtype: int64

In [22]:
coffee.isna().sum().sum()

11989

There are a number of null values that will need to be addressed, particularly around score related data, and lat and lon values. 

**Addressing Missing `Overall_Score`**

There are also a few missing their `overall_score` value, we'll start with those.

In [23]:
overall_boolean = pd.isnull(coffee["overall_score"])
coffee[overall_boolean]

coffee_name                  roaster_name  month  year  \
1593                 Vienna Roast  Alpen Sierra Coffee Roasters     10  1998   
1594  Brazil Fazenda Vista Alegre              City Bean Coffee     10  1998   
1597         Elm City House Blend             Elm City Roasters     10  1998   
1651         Natural Moka – Green       Kaanapali Estate Coffee      4  1998   

      roast  bean_agtron  ground_agtron  aroma  acidity  body  ...  \
1593      4           38             44      7        6     5  ...   
1594      4           45             53      6        5     6  ...   
1597      5           38             43      6        5     7  ...   
1651   <NA>         <NA>           <NA>   <NA>     <NA>  <NA>  ...   

      aftertaste  with_milk  roaster_lat  roaster_lon  origin_lat  origin_lon  \
1593        <NA>       <NA>    39.088540  -120.050353   44.933143    7.540121   
1594        <NA>       <NA>    34.092301  -118.369289   44.933143    7.540121   
1597        <NA>       <NA>    41.308214   -72.925052   44.933143    7.540121   
1651        <NA>       <NA>    20.802957  -156.310683   44.933143    7.540121   

                                                     p1  \
1593  Blind Assessment: Some cups of the sample are ...   
1594  Blind Assessment: Some cups display a muted bu...   
1597  Blind Assessment: About half the cups of this ...   
1651  Blind Assessment: To say this coffee has an at...   

                                                     p2  \
1593  Notes: Not rated due to a taste fault in some ...   
1594  Notes: Not rated due to a taste defect in some...   
1597  Notes: Not rated due to a taste defect called ...   
1651  Notes: This coffee was not rated due to the pa...   

                                                     p3 overall_score  
1593  Who Should Drink It: The best-seller factor:I'...          <NA>  
1594  Who Should Drink It: The best-seller factor:I ...          <NA>  
1597  Who Should Drink It: The best-seller factor:I ...          <NA>  
1651  This review originally appeared in the April, ...          <NA>  

[4 rows x 21 columns]

Investigated these four; they were unrated. Given `overall_score` is the target variable, and only 4 are missing their score, droppiing these four rows. However, we'll save this as a dataframe for later reference just in case.

In [24]:
#saving these 4 as their own df before dropping the rows missing their rating
overall_boolean_df = coffee[overall_boolean]

In [25]:
#drop rows missing value in overall_score column 
coffee.dropna(subset = ['overall_score'], inplace=True)

In [26]:
#checking drop accuracy
coffee.isna().sum()

coffee_name         0
roaster_name        0
month               0
year                0
roast             351
bean_agtron       275
ground_agtron     269
aroma              38
acidity          1065
body               54
flavor           1020
aftertaste       1782
with_milk        5788
roaster_lat        72
roaster_lon        72
origin_lat        589
origin_lon        589
p1                  0
p2                  0
p3                  6
overall_score       0
dtype: int64

**Addressing Missing `with_milk`** 

Next, we'll move on to the column missing the most values `with_milk`. `with_milk` refers to the coffee score when served with milk. Given 1) the proportion of these missing (5788/6563) and 2) the number of other scores we have available, we'll drop this column.

In [27]:
#dropping "with_milk" column
coffee.drop('with_milk', axis=1, inplace=True)

**Addressing Missing `roast`, `bean_agtron`, `ground_agtron`**

`Roast`, `bean_agtron` and `ground_agtron` are all related, as agtron is a numerical range indicating roast level. In this data, `roast` turned text such as `Light` to `Dark` to numerical values 1 through 6.

The bean_agtron and roast_agtron values typically differ slightly as roast is measured both for the bean and once ground. The mean of these is used to then determine roast level. Learn more about the score at [CoffeeReview](https://www.coffeereview.com/roast-definitions/).

To start, we'll see if we can impute values where `bean_agtron` and `ground_agtron` are missing by checking the roast values.

In [28]:
#checking missing values
coffee.isna().sum()

coffee_name         0
roaster_name        0
month               0
year                0
roast             351
bean_agtron       275
ground_agtron     269
aroma              38
acidity          1065
body               54
flavor           1020
aftertaste       1782
roaster_lat        72
roaster_lon        72
origin_lat        589
origin_lon        589
p1                  0
p2                  0
p3                  6
overall_score       0
dtype: int64

In [29]:
#can we get better info to impute by comparing these columns?
coffee.loc[coffee['bean_agtron'].isna(),'roast'].describe()

count    27.000000
mean      1.592593
std       1.009922
min       1.000000
25%       1.000000
50%       1.000000
75%       2.000000
max       5.000000
Name: roast, dtype: float64

It does look like there are some instances where we have the `roast` value but not `bean_agtron`. Updating below using judgment call using coffee review roast guide.

In [30]:
#imputing 75 for bean agtron anywhere roast is 1 
coffee.loc[coffee['bean_agtron'].isna() & (coffee['roast'] == 1),'bean_agtron'] = 75

In [31]:
#imputing 65 for bean agtron anywhere roast is 2
coffee.loc[coffee['bean_agtron'].isna() & (coffee['roast'] == 2),'bean_agtron'] = 65

In [32]:
#imputing 55 for bean agtron anywhere roast is 3
coffee.loc[coffee['bean_agtron'].isna() & (coffee['roast'] == 3),'bean_agtron'] = 55

In [33]:
#imputing 35 for bean agtron anywhere roast is 5
coffee.loc[coffee['bean_agtron'].isna() & (coffee['roast'] == 5),'bean_agtron'] = 35

In [34]:
#can we get better info to impute by comparing these columns?
coffee.loc[coffee['ground_agtron'].isna(),'roast'].describe()

count    21.000000
mean      1.476190
std       1.209093
min       1.000000
25%       1.000000
50%       1.000000
75%       1.000000
max       5.000000
Name: roast, dtype: float64

In [35]:
#imputing 75 for bean agtron anywhere roast is 1 
coffee.loc[coffee['ground_agtron'].isna() & (coffee['roast'] == 1),'ground_agtron'] = 75

In [36]:
#imputing 45 for bean agtron anywhere roast is 4
coffee.loc[coffee['ground_agtron'].isna() & (coffee['roast'] == 4),'ground_agtron'] = 45

In [37]:
#imputing 35 for bean agtron anywhere roast is 5
coffee.loc[coffee['ground_agtron'].isna() & (coffee['roast'] == 5),'ground_agtron'] = 35

In [38]:
#are bean and ground agtron missing in the same spots?
coffee.loc[coffee['ground_agtron'].isna(),'bean_agtron'].describe()

count       0
mean     <NA>
std      <NA>
min      <NA>
25%      <NA>
50%      <NA>
75%      <NA>
max      <NA>
Name: bean_agtron, dtype: object

In [39]:
#are bean and ground agtron missing in the same spots?
coffee.loc[coffee['bean_agtron'].isna(),'ground_agtron'].describe()

count       0
mean     <NA>
std      <NA>
min      <NA>
25%      <NA>
50%      <NA>
75%      <NA>
max      <NA>
Name: ground_agtron, dtype: object

`bean_agtron` and `ground_agtron` are missing values in the same spots.

In [40]:
#digging further into bean_agtron
coffee['ground_agtron'].describe()

count    6315.000000
mean       66.339351
std        15.902328
min         0.000000
25%        58.000000
50%        71.000000
75%        77.000000
max       105.000000
Name: ground_agtron, dtype: float64

Found an odd min for `bean_agtron`. Dug through earlier notebooks and website to see if there might be broader issues with how data was compiled. This appears to be from a website typo.

In [41]:
coffee.loc[coffee['bean_agtron'] == -124]

coffee_name   roaster_name  month  year  roast  bean_agtron  \
894  Look for the Helpers  Tony's Coffee      3  2021      1         -124   

     ground_agtron  aroma  acidity  body  flavor  aftertaste  roaster_lat  \
894             70      8        8     9       8           8    48.754401   

     roaster_lon  origin_lat  origin_lon  \
894  -122.478836    10.21167    38.65212   

                                                    p1  \
894  Blind Assessment: Crisply chocolaty, rich-tone...   

                                                    p2  \
894  Notes: This coffee earned the fourth-highest r...   

                                                    p3  overall_score  
894  The Bottom Line: 100% of the profits from the ...             91

In [42]:
#replacing this value with the correct score
coffee['bean_agtron'].replace(-124, 52, inplace=True)

In [43]:
#checking updated correctly
coffee.loc[coffee['coffee_name'] == 'Look for the Helpers']

coffee_name   roaster_name  month  year  roast  bean_agtron  \
894  Look for the Helpers  Tony's Coffee      3  2021      1           52   

     ground_agtron  aroma  acidity  body  flavor  aftertaste  roaster_lat  \
894             70      8        8     9       8           8    48.754401   

     roaster_lon  origin_lat  origin_lon  \
894  -122.478836    10.21167    38.65212   

                                                    p1  \
894  Blind Assessment: Crisply chocolaty, rich-tone...   

                                                    p2  \
894  Notes: This coffee earned the fourth-highest r...   

                                                    p3  overall_score  
894  The Bottom Line: 100% of the profits from the ...             91

In [44]:
#rechecking out distribution of bean_agtron
coffee['bean_agtron'].describe()

count    6315.000000
mean       49.401267
std        14.406642
min         0.000000
25%        46.000000
50%        53.000000
75%        58.000000
max        86.000000
Name: bean_agtron, dtype: float64

Based on agtron meaning, zero seems like an odd value to have for so many (essentially would be completely burnt). Checking if we have zero for both `bean_agtron` and `ground_agtron` in same places.

In [45]:
(coffee['bean_agtron'] == 0).sum()

346

In [46]:
(coffee['ground_agtron'] == 0).sum()

112

In [47]:
#seeing if we can impute bean_agtron using ground_agtron
#does appear some of these 0s must be mistakes
coffee.loc[(coffee['bean_agtron'] == 0) & (coffee['ground_agtron'] != 0)]

coffee_name           roaster_name  month  \
2157                 The Knight Dark Roast             Glory Brew     11   
2595     SFCC House Blend Espresso Capsule              Gourmesso      7   
2657               O’Rey Espresso Capsules           Lino'spresso      5   
2676               Breakfast Blend (K-Cup)  Green Mountain Coffee      5   
2714  Organic Ethiopia Yirgacheffe (K-Cup)  Green Mountain Coffee      3   
...                                    ...                    ...    ...   
1323                      Costa Rica Roast          Café Buen Dia      3   
1324                            Mild Roast            Hills Bros.      3   
1325                            Dark Roast           Café Bustelo      3   
1326                         Classic Roast               Folgers       3   
1327                          Master Blend          Maxwell House      3   

      year  roast  bean_agtron  ground_agtron  aroma  acidity  body  flavor  \
2157  2017      6            0             37      8        7     9       8   
2595  2016      6            0             58      7     <NA>     8       8   
2657  2016      6            0             50      8     <NA>     7       7   
2676  2016      6            0             56      8        7     7       8   
2714  2016      6            0             48      8        8     8       9   
...    ...    ...          ...            ...    ...      ...   ...     ...   
1323  2002      6            0             38      5        5     7       5   
1324  2002      6            0             67      5        5     6       5   
1325  2002      6            0             44      5        1     8       5   
1326  2002      6            0             64      6        5     7       5   
1327  2002      6            0             63      5     <NA>     7       6   

      aftertaste  roaster_lat  roaster_lon  origin_lat  origin_lon  \
2157           7    25.774173   -80.193620   -2.483383  117.890285   
2595           7    52.517037    13.388860  -30.292848  153.125616   
2657           7    34.053691  -118.242766   21.417531   39.886174   
2676           7    44.337125   -72.755497         NaN         NaN   
2714           8    44.337125   -72.755497   10.211670   38.652120   
...          ...          ...          ...         ...         ...   
1323        <NA>    33.703815  -117.962735   44.933143    7.540121   
1324        <NA>          NaN          NaN   44.933143    7.540121   
1325        <NA>    25.774173   -80.193620   44.933143    7.540121   
1326        <NA>    39.102797   -84.506462   44.933143    7.540121   
1327        <NA>          NaN          NaN   44.933143    7.540121   

                                                     p1  \
2157  Blind Assessment: (As brewed in a Keurig 2.0 K...   
2595  Blind Assessment: Evaluated as espresso produc...   
2657  Blind Assessment: Evaluated as espresso. Lean,...   
2676  Blind Assessment: (As brewed in a Keurig 2.0 K...   
2714  Blind Assessment: (As brewed in a Keurig 2.0 K...   
...                                                 ...   
1323  Blind Assessment: Low-key to the point of neut...   
1324  Blind Assessment: The aroma is thin, faintly s...   
1325  Blind Assessment: Densely bitter; heavy rather...   
1326  Blind Assessment: Dominated by sweet grainy nu...   
1327  Blind Assessment: Cloyingly sweet and sharply ...   

                                                     p2  \
2157  Notes: A blend of coffees from Indonesia and C...   
2595  Notes: This coffee is Fair Trade certified, me...   
2657  Notes: Lino'spresso produces authentic Italian...   
2676  Notes: Keurig/Green Mountain, formerly Green M...   
2714  Notes: This coffee is certified organically gr...   
...                                                 ...   
1323  Notes: Sold roast and ground in a 12-ounce val...   
1324  Notes: Sold roasted and ground in an 11.5-ounc...   
1325  Notes: Sold roast and ground in a 10-ounce can...   
1326  Notes: Sold roast and g

In [48]:
imputed_agtron = coffee.loc[(coffee['bean_agtron'] == 0) & (coffee['ground_agtron'] != 0), 'ground_agtron']

In [49]:
len(coffee.loc[coffee['bean_agtron'] == 0])

346

In [50]:
imputed_agtron

2157    37
2595    58
2657    50
2676    56
2714    48
        ..
1323    38
1324    67
1325    44
1326    64
1327    63
Name: ground_agtron, Length: 243, dtype: Int8

In [51]:
#where bean_agtron is zero but ground agtron is not, impute bean_agtron with ground_agtron score
coffee.loc[(coffee['bean_agtron'] == 0) & (coffee['ground_agtron'] != 0), 'bean_agtron'] = imputed_agtron

In [52]:
#checking values updated where expected
(coffee['bean_agtron'] == 0).sum()

103

In [53]:
len(coffee.loc[(coffee['ground_agtron'] == 0)])

112

In [54]:
imputed_agtron_2 = coffee.loc[(coffee['ground_agtron'] == 0) & (coffee['bean_agtron'] != 0), 'bean_agtron']

In [55]:
len(imputed_agtron_2)

9

In [56]:
#where ground_agtron is zero but bean agtron is not, impute ground_agtron with bean_agtron score
coffee.loc[(coffee['ground_agtron'] == 0) & (coffee['bean_agtron'] != 0), 'ground_agtron'] = imputed_agtron_2

In [57]:
#verifying updates appear as expected
len(coffee.loc[(coffee['ground_agtron'] == 0)])

103

We've now imputed agtron values using related columsn where possible. We still need to address:
- Agtron scores of zero
- Missing agtron values

Addressing agtron scores of zero:

In [58]:
#checking central values we can use to impute
#mean and median aren't too far off, going with median since data seems slightly skewed
coffee['bean_agtron'].describe()

count    6315.000000
mean       51.287411
std        10.816049
min         0.000000
25%        47.000000
50%        54.000000
75%        58.000000
max        91.000000
Name: bean_agtron, dtype: float64

In [59]:
#where bean_agtron is zero impute with median
coffee.loc[(coffee['bean_agtron'] == 0), 'bean_agtron'] = 54

In [60]:
#addressing ground agtron
coffee['ground_agtron'].describe()

count    6315.000000
mean       66.408234
std        15.721128
min         0.000000
25%        58.000000
50%        71.000000
75%        77.000000
max       105.000000
Name: ground_agtron, dtype: float64

In [61]:
#where ground_agtron is zero impute with median
coffee.loc[(coffee['ground_agtron'] == 0), 'ground_agtron'] = 71

In [62]:
#addressing nan bean_agtron
coffee.loc[(coffee['bean_agtron'].isna()), 'bean_agtron'] = 54

In [63]:
#where bean_agtron is zero impute with median
coffee.loc[(coffee['ground_agtron'].isna()), 'ground_agtron'] = 71

In [64]:
#verifying no nans left for bean and ground agtron
coffee.isna().sum()

coffee_name         0
roaster_name        0
month               0
year                0
roast             351
bean_agtron         0
ground_agtron       0
aroma              38
acidity          1065
body               54
flavor           1020
aftertaste       1782
roaster_lat        72
roaster_lon        72
origin_lat        589
origin_lon        589
p1                  0
p2                  0
p3                  6
overall_score       0
dtype: int64

Because `roast`,`bean_agtron` and `ground_agrton` are measuring similar things but `bean_agtron` and `ground_agrton` add a level of complexity, we'll drop roast for now. 

In [65]:
#dropping "roast" column
coffee.drop('roast', axis=1, inplace=True)

**Addressing missing score values**

From initial experience with the data, we suspect the aroma, acidity, body, flavor, and aftertaste scores to be highly correlated and potentially useful for imputing.

`Aroma` only has a few missing, so we'll beging with that value.

In [66]:
corr_aroma = coffee.corr()

corr_aroma.loc[:,"aroma"].sort_values(ascending = False)

aroma            1.000000
overall_score    0.836090
flavor           0.779697
acidity          0.718280
body             0.615699
year             0.613025
aftertaste       0.482069
ground_agtron    0.425609
bean_agtron      0.312716
roaster_lon      0.164530
month            0.097422
origin_lon      -0.011369
roaster_lat     -0.051370
origin_lat      -0.408554
Name: aroma, dtype: float64

`aroma` is highly correlated with `overall_score`, but given that is our target value. Instead we'll use flavor, which it is also highly correlated with.

In [67]:
#checking out flavor composition
coffee['flavor'].describe()

count    5543.000000
mean        8.548259
std         1.012805
min         1.000000
25%         8.000000
50%         9.000000
75%         9.000000
max        10.000000
Name: flavor, dtype: float64

In [68]:
#calculate the constant of proportionality, we'll use mean given how 
constant = (coffee['aroma']/coffee['flavor']).mean()
constant

0.9952009063705389

In [69]:
#calcuate aroma values to impute
aroma_values = round(constant*coffee.loc[coffee['aroma'].isna(),'flavor']).astype('Int8')
aroma_values

113        9
1261       9
1262       9
1263       9
1264       9
1265       9
1266       9
1267       9
1268       9
1269       9
1270       9
1271       9
1929       9
1966       9
1967       9
1968       9
1969       9
1970       9
1971       9
1972       9
1973       9
1974       9
1975       8
1976       9
2426       9
2427       9
2429       9
3307       9
3308       9
3309       9
3310       9
3311       8
3312       8
3313       8
3314       8
3479       8
1150    <NA>
1371       9
Name: flavor, dtype: Int8

In [70]:
#impute values based on correlations
coffee.loc[coffee['aroma'].isna(),'aroma'] = aroma_values

In [71]:
coffee['aroma'].describe()

count    6562.000000
mean        8.361475
std         0.927207
min         2.000000
25%         8.000000
50%         9.000000
75%         9.000000
max        10.000000
Name: aroma, dtype: float64

In [72]:
#update last missing nan with rounded aroma mean
coffee.loc[coffee['aroma'].isna(),'aroma'] = 8

Addressing `acidity` missing values:

In [73]:
corr_acidity = coffee.corr()

corr_acidity.loc[:,"acidity"].sort_values(ascending = False)

acidity          1.000000
overall_score    0.862590
flavor           0.813859
aroma            0.717851
year             0.660837
body             0.653796
aftertaste       0.584571
ground_agtron    0.509627
bean_agtron      0.405224
roaster_lon      0.176663
month            0.101833
origin_lon      -0.031112
roaster_lat     -0.046584
origin_lat      -0.482881
Name: acidity, dtype: float64

Again, highly correlated with target `overall_score` but we'll work with `flavor`, also closely correlated.

In [74]:
#calculate the constant of proportionality, we'll use mean given how 
constant = (coffee['acidity']/coffee['flavor']).mean()

In [75]:
#calcuate values to impute
acidity_values = round(constant*coffee.loc[coffee['acidity'].isna(),'flavor']).astype('Int8')
acidity_values

11         8
15         8
19         8
30         8
31         8
        ... 
1717       4
1718       4
1719    <NA>
1720       6
1721       7
Name: flavor, Length: 1065, dtype: Int8

In [76]:
#impute values based on correlations
coffee.loc[coffee['acidity'].isna(),'acidity'] = acidity_values

In [77]:
coffee['acidity'].describe()

count    6542.000000
mean        7.904311
std         1.024400
min         1.000000
25%         8.000000
50%         8.000000
75%         9.000000
max        10.000000
Name: acidity, dtype: float64

In [78]:
#update last missing nan with acidity rounded average(happens to be same as the median)
coffee.loc[coffee['acidity'].isna(),'acidity'] = 8

Addressing missing `body` values:

In [79]:
corr_body = coffee.corr()

corr_body.loc[:,"body"].sort_values(ascending = False)

body             1.000000
overall_score    0.743217
year             0.693208
flavor           0.681260
acidity          0.633459
aroma            0.615373
ground_agtron    0.446841
aftertaste       0.353036
bean_agtron      0.350251
roaster_lon      0.183678
month            0.079199
origin_lon      -0.038704
roaster_lat     -0.068673
origin_lat      -0.405930
Name: body, dtype: float64

Seems odd for it to be highly correlated with year, but could work. Since `flavor` is so close, we'll stick with that.

In [80]:
#calculate the constant of proportionality, we'll use mean given how 
constant = (coffee['body']/coffee['flavor']).mean()
constant

0.9694178128284203

In [81]:
#calcuate values to impute
body_values = round(coffee.loc[coffee['body'].isna(),'flavor']).astype('Int8')
body_values

6       <NA>
62      <NA>
73      <NA>
152     <NA>
153     <NA>
154     <NA>
210     <NA>
219     <NA>
223     <NA>
277     <NA>
278     <NA>
325     <NA>
327     <NA>
335     <NA>
381     <NA>
382     <NA>
384     <NA>
402     <NA>
410     <NA>
422     <NA>
424     <NA>
429     <NA>
477     <NA>
518     <NA>
526     <NA>
578     <NA>
579     <NA>
586     <NA>
589     <NA>
655     <NA>
675     <NA>
734     <NA>
798     <NA>
801     <NA>
818     <NA>
822     <NA>
890     <NA>
907     <NA>
909     <NA>
931     <NA>
934     <NA>
1059    <NA>
1070    <NA>
1071    <NA>
1150    <NA>
1269    <NA>
1270    <NA>
1380    <NA>
1494    <NA>
1590    <NA>
1605    <NA>
1614    <NA>
1685    <NA>
1772    <NA>
Name: flavor, dtype: Int8

In [82]:
#are body and flavor missing in the same spots?
coffee.loc[coffee['body'].isna(),'flavor'].describe()

count       0
mean     <NA>
std      <NA>
min      <NA>
25%      <NA>
50%      <NA>
75%      <NA>
max      <NA>
Name: flavor, dtype: object

Body and flavor are missing in the same spots, let's try a different value to reference:

In [83]:
#calculate the constant of proportionality, we'll use mean given how 
constant = (coffee['body']/coffee['acidity']).mean()
constant

1.0387045897377742

In [84]:
#calcuate aroma values to impute
body_values = round(coffee.loc[coffee['body'].isna(),'acidity']).astype('Int8')
body_values

6       8
62      8
73      8
152     8
153     8
154     8
210     8
219     7
223     8
277     7
278     7
325     8
327     8
335     8
381     7
382     7
384     7
402     8
410     8
422     8
424     8
429     8
477     7
518     8
526     7
578     7
579     8
586     8
589     9
655     8
675     7
734     8
798     7
801     8
818     7
822     9
890     7
907     7
909     8
931     8
934     8
1059    6
1070    7
1071    7
1150    8
1269    7
1270    7
1380    7
1494    6
1590    7
1605    7
1614    8
1685    5
1772    8
Name: acidity, dtype: Int8

In [85]:
#impute values based on correlations
coffee.loc[coffee['body'].isna(),'body'] = body_values

In [86]:
#checking where there are still missing values
coffee.isna().sum()

coffee_name         0
roaster_name        0
month               0
year                0
bean_agtron         0
ground_agtron       0
aroma               0
acidity             0
body                0
flavor           1020
aftertaste       1782
roaster_lat        72
roaster_lon        72
origin_lat        589
origin_lon        589
p1                  0
p2                  0
p3                  6
overall_score       0
dtype: int64

Addressing `flavor` missing values:

In [87]:
corr_flavor = coffee.corr()

corr_flavor.loc[:,"flavor"].sort_values(ascending = False)

flavor           1.000000
overall_score    0.920572
acidity          0.815014
aroma            0.780039
body             0.681260
year             0.657944
aftertaste       0.595812
ground_agtron    0.462309
bean_agtron      0.351741
roaster_lon      0.141835
month            0.102250
roaster_lat     -0.043239
origin_lon      -0.051695
origin_lat      -0.549698
Name: flavor, dtype: float64

In [88]:
#are these missing in the same spots?
coffee.loc[coffee['flavor'].isna(),'acidity'].describe()

count    1020.000000
mean        7.350000
std         0.718499
min         3.000000
25%         7.000000
50%         7.000000
75%         8.000000
max        10.000000
Name: acidity, dtype: float64

In [89]:
#calculate the constant of proportionality, we'll use mean given how 
constant = (coffee['flavor']/coffee['acidity']).mean()
constant

1.0757197695362946

In [90]:
#calcuate aroma values to impute
flavor_values = round(coffee.loc[coffee['flavor'].isna(),'acidity']).astype('Int8')
flavor_values

0       8
1       8
2       8
3       9
4       8
       ..
1772    8
1780    6
1781    7
1786    6
1787    4
Name: acidity, Length: 1020, dtype: Int8

In [91]:
#impute values based on correlations
coffee.loc[coffee['flavor'].isna(),'flavor'] = flavor_values

In [92]:
#checking where there are still missing values
coffee.isna().sum()

coffee_name         0
roaster_name        0
month               0
year                0
bean_agtron         0
ground_agtron       0
aroma               0
acidity             0
body                0
flavor              0
aftertaste       1782
roaster_lat        72
roaster_lon        72
origin_lat        589
origin_lon        589
p1                  0
p2                  0
p3                  6
overall_score       0
dtype: int64

In [93]:
#these aren't as strongly correlated, but we'll work with it still for now using flavor
corr_flavor = coffee.corr()

corr_flavor.loc[:,"aftertaste"].sort_values(ascending = False)

aftertaste       1.000000
overall_score    0.764557
flavor           0.595812
acidity          0.544620
aroma            0.481441
body             0.353036
ground_agtron    0.276069
bean_agtron      0.188278
year             0.101309
month            0.052519
roaster_lon      0.039686
origin_lon       0.021183
origin_lat      -0.027026
roaster_lat     -0.042939
Name: aftertaste, dtype: float64

In [94]:
#calculate the constant of proportionality, we'll use mean given how 
constant = (coffee['aftertaste']/coffee['flavor']).mean()
constant

0.9142936823338413

In [95]:
#calcuate aroma values to impute
after_values = round(coffee.loc[coffee['aftertaste'].isna(),'flavor']).astype('Int8')
after_values

0       8
1       8
2       8
3       9
4       8
       ..
1783    5
1784    7
1785    8
1786    6
1787    4
Name: flavor, Length: 1782, dtype: Int8

In [96]:
#impute values based on correlations
coffee.loc[coffee['aftertaste'].isna(),'aftertaste'] = after_values

Checking back in on nans - getting close:

In [97]:
#checking where there are still missing values
coffee.isna().sum()

coffee_name        0
roaster_name       0
month              0
year               0
bean_agtron        0
ground_agtron      0
aroma              0
acidity            0
body               0
flavor             0
aftertaste         0
roaster_lat       72
roaster_lon       72
origin_lat       589
origin_lon       589
p1                 0
p2                 0
p3                 6
overall_score      0
dtype: int64

**Addressing missing `p3` values**

`p3` refers to the third scraped paragraph. Since there are so few rows missing this, we'll drop those rows.

In [98]:
#drop rows missing value in overall_score column 
coffee.dropna(subset = ['p3'], inplace=True)

**Addressing missing latitude and longitude values**

We are missing some lat and lon values. First we'll confirm the lat and lon are missing in the same places:

In [99]:
#are these missing in the same spots?
coffee.loc[coffee['roaster_lat'].isna(),'roaster_lon'].describe()

count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: roaster_lon, dtype: float64

In [100]:
#are these missing in the same spots?
coffee.loc[coffee['origin_lat'].isna(),'origin_lon'].describe()

count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: origin_lon, dtype: float64

We don't have great information to try to work from for imputing the lat and lon in meaningful ways. For now, we'll use the mean for these so that we can retain the other column values in these spots.

In [101]:
coffee['origin_lat'].describe()

count    5968.000000
mean       10.198811
std        13.321592
min       -30.292848
25%         1.441968
50%        10.211670
75%        14.897192
max        52.380033
Name: origin_lat, dtype: float64

In [102]:
coffee['origin_lon'].describe()

count    5968.000000
mean      -12.161417
std        72.166777
min      -155.428370
25%       -81.130843
50%         7.540121
75%        38.652120
max       153.125616
Name: origin_lon, dtype: float64

In [103]:
coffee['roaster_lat'].describe()

count    6485.000000
mean       37.027419
std         9.556831
min       -37.824425
25%        32.776272
50%        39.049011
75%        43.447436
max        64.837845
Name: roaster_lat, dtype: float64

In [104]:
coffee['roaster_lon'].describe()

count    6485.000000
mean      -64.412448
std        82.161370
min      -159.582761
25%      -117.648430
50%       -89.383761
75%       -73.121221
max       152.952215
Name: roaster_lon, dtype: float64

In [105]:
#impute values based on mean
coffee.loc[coffee['origin_lat'].isna(),'origin_lat'] = 10.198811

In [106]:
#impute values based on mean
coffee.loc[coffee['origin_lon'].isna(),'origin_lon'] = -12.161417

In [107]:
#impute values based on mean
coffee.loc[coffee['roaster_lat'].isna(),'roaster_lat'] = 37.027489

In [108]:
#impute values based on mean
coffee.loc[coffee['roaster_lon'].isna(),'roaster_lon'] = -64.412406

In [109]:
#checking where there are still missing values
coffee.isna().sum()

coffee_name      0
roaster_name     0
month            0
year             0
bean_agtron      0
ground_agtron    0
aroma            0
acidity          0
body             0
flavor           0
aftertaste       0
roaster_lat      0
roaster_lon      0
origin_lat       0
origin_lon       0
p1               0
p2               0
p3               0
overall_score    0
dtype: int64

### Checking numerical data for unusual values, initial insights

In [110]:
#checking for any other unusual values
coffee.describe()

month         year  bean_agtron  ground_agtron        aroma  \
count  6557.000000  6557.000000  6557.000000    6557.000000  6557.000000   
mean      6.642062  2013.102181    52.229983      67.683697     8.360988   
std       3.437202     6.719059     8.411739      12.962223     0.927393   
min       1.000000  1997.000000    11.000000      12.000000     2.000000   
25%       4.000000  2008.000000    48.000000      60.000000     8.000000   
50%       7.000000  2014.000000    54.000000      71.000000     9.000000   
75%      10.000000  2019.000000    58.000000      77.000000     9.000000   
max      12.000000  2022.000000    91.000000     105.000000    10.000000   

           acidity         body       flavor   aftertaste  roaster_lat  \
count  6557.000000  6557.000000  6557.000000  6557.000000  6557.000000   
mean      7.904224     8.086472     8.361598     7.795638    37.027419   
std       1.023060     0.871203     1.065692     0.874365     9.504209   
min       1.000000     4.000000     1.000000     2.000000   -37.824425   
25%       8.000000     8.000000     8.000000     7.000000    32.840162   
50%       8.000000     8.000000     9.000000     8.000000    39.049011   
75%       9.000000     9.000000     9.000000     8.000000    43.074761   
max      10.000000    10.000000    10.000000    10.000000    64.837845   

       roaster_lon   origin_lat   origin_lon  overall_score  
count  6557.000000  6557.000000  6557.000000    6557.000000  
mean    -64.412448    10.198811   -12.161417      90.873723  
std      81.708964    12.709097    68.848718       4.089739  
min    -159.582761   -30.292848  -155.428370      60.000000  
25%    -117.423510     1.441968   -79.366697      90.000000  
50%     -89.383761    10.211670     7.540121      92.000000  
75%     -72.815099    13.800038    38.652120      93.000000  
max     152.952215    52.380033   153.125616      98.000000

**Things of note from description:**
This looks pretty good because we address nans and odd zero values above.

    - 105 for ground agtron seems high, dig further
    - score columns seem skewed towards 10, may need to adjust to even out distribution for these
    - overall score mean is 90 (high end of range; scores can go from 50-100, in these set range from 60 to 98)
    
We'll address potential skew and high `overall_score` means in a future notebook.

**Addressing suspiciously high `ground_agtron`**

In [111]:
#investigating spots with high groud_agtron
coffee.loc[coffee['ground_agtron'] >= 100]

coffee_name  \
682                            Brazil Gauriroba Natural   
886                           Colombia Cauca El Paraiso   
3055  Nicaragua Finca La Joya Villa Lobos Natural Sa...   
3174                             Perci Red Panama Gesha   

                        roaster_name  month  year  bean_agtron  ground_agtron  \
682               SOT Coffee Roaster      7  2021           82            104   
886               SOT Coffee Roaster      3  2021           85            105   
3055                   Big Nose Cafe      4  2015           72            102   
3174  Barrington Coffee Roasting Co.     12  2014           74            101   

      aroma  acidity  body  flavor  aftertaste  roaster_lat  roaster_lon  \
682       9        9     9       9           8    34.619881   135.490357   
886       9        9     9      10           9    34.619881   135.490357   
3055      9        8     8       9           8    24.816472   120.970516   
3174      9        9     9       9           9    42.303920   -73.234097   

      origin_lat  origin_lon  \
682   -10.333333  -53.200000   
886     4.099917  -72.908813   
3055   12.609016  -85.293691   
3174    8.559559  -81.130843   

                                                     p1  \
682   Blind Assessment: Citrusy-bright, juicy-sweet,...   
886   Blind Assessment: Lyrically sweet, juicy, eleg...   
3055  Blind Assessment: Sweetly and richly fermenty:...   
3174  Blind Assessment: Impressive balance and compl...   

                                                     p2  \
682   Notes: Produced by Homero Aguiar Paiva at Faze...   
886   Notes: SOT Coffee Roaster, based in Osaka, Jap...   
3055  Notes: This is a dry or “natural” processed co...   
3174  Notes: Produced from trees of the rare Ethiopi...   

                                                     p3  overall_score  
682   The Bottom Line: An intriguingly complex Brazi...             94  
886   The Bottom Line: An exquisitely refined ultra-...             96  
3055  Who Should Drink It: An unapologetic, thorough...             92  
3174  Who Should Drink It: A very light-roasted pres...             95

There are a few `ground_agtron` spots where `bean_agtron` is much lower. Will impute `ground_agtron` with corresponding `bean_agtron` values so it is not skewing our data.

In [112]:
#impute values based on mean
coffee.loc[coffee['ground_agtron'] == 105,'ground_agtron'] = 85

In [113]:
#impute values based on mean
coffee.loc[coffee['ground_agtron'] == 104,'ground_agtron'] = 82

In [114]:
#impute values based on mean
coffee.loc[coffee['ground_agtron'] == 102,'ground_agtron'] = 72

In [115]:
#impute values based on mean
coffee.loc[coffee['ground_agtron'] == 101,'ground_agtron'] = 74

In [116]:
#checking for any other unusual values
coffee.describe()

month         year  bean_agtron  ground_agtron        aroma  \
count  6557.000000  6557.000000  6557.000000    6557.000000  6557.000000   
mean      6.642062  2013.102181    52.229983      67.668598     8.360988   
std       3.437202     6.719059     8.411739      12.936090     0.927393   
min       1.000000  1997.000000    11.000000      12.000000     2.000000   
25%       4.000000  2008.000000    48.000000      60.000000     8.000000   
50%       7.000000  2014.000000    54.000000      71.000000     9.000000   
75%      10.000000  2019.000000    58.000000      77.000000     9.000000   
max      12.000000  2022.000000    91.000000      99.000000    10.000000   

           acidity         body       flavor   aftertaste  roaster_lat  \
count  6557.000000  6557.000000  6557.000000  6557.000000  6557.000000   
mean      7.904224     8.086472     8.361598     7.795638    37.027419   
std       1.023060     0.871203     1.065692     0.874365     9.504209   
min       1.000000     4.000000     1.000000     2.000000   -37.824425   
25%       8.000000     8.000000     8.000000     7.000000    32.840162   
50%       8.000000     8.000000     9.000000     8.000000    39.049011   
75%       9.000000     9.000000     9.000000     8.000000    43.074761   
max      10.000000    10.000000    10.000000    10.000000    64.837845   

       roaster_lon   origin_lat   origin_lon  overall_score  
count  6557.000000  6557.000000  6557.000000    6557.000000  
mean    -64.412448    10.198811   -12.161417      90.873723  
std      81.708964    12.709097    68.848718       4.089739  
min    -159.582761   -30.292848  -155.428370      60.000000  
25%    -117.423510     1.441968   -79.366697      90.000000  
50%     -89.383761    10.211670     7.540121      92.000000  
75%     -72.815099    13.800038    38.652120      93.000000  
max     152.952215    52.380033   153.125616      98.000000

In [117]:
coffee.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6557 entries, 0 to 1787
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   coffee_name    6557 non-null   object 
 1   roaster_name   6557 non-null   object 
 2   month          6557 non-null   int64  
 3   year           6557 non-null   int64  
 4   bean_agtron    6557 non-null   Int8   
 5   ground_agtron  6557 non-null   Int8   
 6   aroma          6557 non-null   Int8   
 7   acidity        6557 non-null   Int8   
 8   body           6557 non-null   Int8   
 9   flavor         6557 non-null   Int8   
 10  aftertaste     6557 non-null   Int8   
 11  roaster_lat    6557 non-null   float64
 12  roaster_lon    6557 non-null   float64
 13  origin_lat     6557 non-null   float64
 14  origin_lon     6557 non-null   float64
 15  p1             6557 non-null   object 
 16  p2             6557 non-null   object 
 17  p3             6557 non-null   object 
 18  overall_

**Still to address: What to do with coffee_name, roaster_name, paragraphs**

In [118]:
#saving cleaned data to CSV
coffee.to_csv('coffee.csv', index=False)